### PR TITLE
treewide: makeWrapper buildInputs to nativeBuildInputs, makeWrapper: use proper shell when cross-compiling

### PR DESCRIPTION
**Pull request by siraben**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by siraben** [comment]
Even though I notice potentially many other incorrect placements of dependencies in `buildInputs`, I will restrict myself to moving `makeWrapper` only.

**Comment by siraben** [comment]
This is being done manually. Many of the packages that I moved `makeWrapper` into `nativeBuildInputs` use it in the `installPhase`. So it should not actually affect the end result.

**Comment by yorickvP** [comment]
Last time I checked, `makeWrapper` only behaved correctly in`buildInputs`, but not `nativeBuildInputs`. Did this get fixed?

**Comment by yorickvP** [comment]
It did not get fixed and this PR breaks cross-compilation for all these packages :D. The offending code is at https://github.com/nixos/nixpkgs/blob/master/pkgs/top-level/all-packages.nix#L513..L514

**Comment by xaverdh** [comment]
@yorickvP can you elaborate on how exactly this breaks cross?

**Comment by yorickvP** [comment]
The makeWrapper from buildPackages makes a wrapper with the shell from the builder, not the target.

**Comment by siraben** [comment]
@Ericson2314 could you help here?

**Comment by Ericson2314** [comment]
Yeah use `targetPackages.runtimeShell`

**Comment by cole-h** [comment]
@ofborg eval

**Comment by siraben** [comment]
Any updates?

**Comment by SuperSandro2000** [comment]
> The makeWrapper from buildPackages makes a wrapper with the shell from the builder, not the target.

Is that fixed or not?

> Yeah use shell = targetPackages.runtimeShell in the definition.

Does this still need to be done?

**Comment by siraben** [comment]
> > Yeah use shell = targetPackages.runtimeShell in the definition.
> 
> 
> 
> Does this still need to be done?

I have a commit in this PR that does it.

**Comment by siraben** [comment]
Rebased.

File: pkgs/development/tools/database/sqldeveloper/default.nix [hunk 1]
```diff
@@ -49,7 +49,7 @@ in
     sha256 = "1h53gl41ydr7kim6q9ckg3xyhb0rhmwj7jnis0xz6vms52b3h59k";
   };
 
-  buildInputs = [ makeWrapper unzip ];
+  nativeBuildInputs = [ makeWrapper unzip ];
```
[the remark this answers is not in the document]
Comment (siraben) [reply pkgs/development/tools/database/sqldeveloper/default.nix:52]: Ok I will undo those types of changes.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 112276` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>hydra-unstable: log was empty</li>
  </ul>
</details>
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>home-assistant</li>
<li>nixpkgs-review</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
hydra-unstable:
<details>


`gpl3` is a deprecated license, check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/development/tools/misc/hydra/common.nix:133:5:

```
    |
133 |     license = licenses.gpl3;
    |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
autoconf is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
automake is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
libtool is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
makeWrapper is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
perl is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
subversion is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
unzip is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/misc/hydra/common.nix:81:3:

```
   |
81 |   buildInputs =
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
Unused argument: `fetchFromGitHub`.
Near pkgs/development/tools/misc/hydra/common.nix:1:40:

```
  |
1 | { stdenv, nix, perlPackages, buildEnv, fetchFromGitHub
  |                                        ^
```
Unused argument: `darcs`.
Near pkgs/development/tools/misc/hydra/common.nix:3:23:

```
  |
3 | , top-git, mercurial, darcs, subversion, breezy, openssl, bzip2, libxslt
  |                       ^
```
Unused argument: `guile`.
Near pkgs/development/tools/misc/hydra/common.nix:4:3:

```
  |
4 | , guile, perl, postgresql, nukeReferences, git, boehmgc, nlohmann_json
  |   ^
```
</details>
</details>

**Review by SuperSandro2000 — APPROVED** [review]
We should do another PR where we move unzip to native.

**Comment by siraben** [comment]
> We should do another PR where we move unzip to native.

see #112302

**Comment by thefloweringash** [comment]
Thanks for cleaning things up :+1:

FYI this broke `elasticsearch6` and `elasticsearch7` with `enableUnfree = true`. They had a trap of overwriting `nativeBuildInputs` to add `autoPatchelfHook`, which removes `makeWrapper` and fails with:

```
installing
/nix/store/zf4mhdlksqlfnqi2psdkfpcdbm56jw1f-stdenv-darwin/setup: line 1321: wrapProgram: command not found
builder for '/nix/store/1ky7sybw853746gxcl3f64ybs2y4gjdx-elasticsearch-7.5.1.drv' failed with exit code 127
```

**Comment by r-burns** [comment]
Also, it looks like commit https://github.com/NixOS/nixpkgs/pull/112276/commits/e8993319724901f132dd4ed997f5383c4a482940 broke evaluation for cross android stdenv, by causing infinite recursion.

Can be reproduced with:
`nix-instantiate -A pkgsCross.aarch64-android-prebuilt.stdenv`
`nix-instantiate -A pkgsCross.armv7a-android-prebuilt.stdenv`

**Comment by lopsided98** [comment]
This breaks cross-compiled `python3.buildEnv`/`python3.withPackages` with an eval failure:
```
error: --- EvalError --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
at: (520:79) in file: /home/ben/Documents/Projects/nixpkgs/pkgs/top-level/all-packages.nix

   519| 
   520|   makeWrapper = makeSetupHook { deps = [ dieHook ]; substitutions = { shell = targetPackages.runtimeShell; }; }
      |                                                                               ^
   521|                               ../build-support/setup-hooks/make-wrapper.sh;

attribute 'runtimeShell' missing
```

**Comment by samueldr** [comment]
```
.../nixpkgs $ env -i nix-build -A pkgsCross.aarch64-multiplatform.python3Packages.libevdev
error: attribute 'runtimeShell' missing, at .../nixpkgs/pkgs/top-level/all-packages.nix:517:79
(use '--show-trace' to show detailed location information)
```

Reverting e8993319724901f132dd4ed997f5383c4a482940 fixes eval.

It seems to affect anything using `buildPythonPackage`.

This happens at the merge commit of this PR (c34b032ccb5cb1cf282ed141d7f60420879e2b42), on the current tip of nixos-unstable (0aeba64fb26e4defa0842a942757144659c6e29f), and at the tip of staging (d39fafe40933de071535f28084b8503bfa9204b8). Note that #114902 changes nothing to this issue.

File: pkgs/applications/editors/vim/macvim-configurable.nix [hunk 2]
```diff
@@ -29,7 +29,7 @@ let
           "/Applications/MacVim.app/Contents/MacOS"
           "/Applications/MacVim.app/Contents/bin"
         ];
-        buildInputs = [ makeWrapper ];
+        nativeBuildInputs = [ makeWrapper ];
```
Comment (lilyball) [inline pkgs/applications/editors/vim/macvim-configurable.nix:32]: This is a `buildEnv` (which doesn't support `nativeBuildInputs`). I can submit a PR to revert the change in this file, but I wonder if other packages were broken this way as well, and whether `buildEnv` should perhaps support `nativeBuildInputs` after all?

**Pull request merged**